source/common/http/http2/codec_impl.cc

```diff
@@ -537,7 +537,16 @@ void ConnectionImpl::ServerStreamImpl::resetStream(StreamResetReason reason) {
     buffer_memory_account_->clearDownstream();
   }
 
-  StreamImpl::resetStream(reason);
+  if (buffer_memory_account_ && buffer_memory_account_->envoyStreamComplete()) {
+    // We should use a lower level reset mechanism to reset the stream.
+    resetStreamWorker(reason);
+    if (parent_.sendPendingFramesAndHandleError()) {
```

Comment:
After some more offline discussion I'd suggested using local_end_stream_ rather than adding yet another way to communicate a stream is done.  @KBaichoo thinks that's sketchy because of interplay between local_end_stream_ and the deferred reset code.
Pulling in @yanavlasov and @antoniovicente to comment.

Reply:
interplay between local_end_stream and `runResetCallbacks(reason)` I think can be sketchy (sorry if I didn't convey that well)

Reply:
This change feels brittle.

I think we need to dig deeper and make object lifetimes more clear.  We need to eliminate all cases where an object remains registered after deletion.

Reply:
yeah that's my thought too.  @KBaichoo if we're in a hurry to get this landed I'd be OK merging with a TODO to fix if you can help resolve the real lifetime issues right after, but if we're not in a rush I think we want to sort out the lifetime issues now in a less fragile way.
/wait

Reply:
Thanks for the pushback. I've dug deeper and have done the following:
* Delay H1 ActiveRequest dtor until after the Envoy level stream is destroyed -- didn't have the invariant that codec outlived the stream in local reply cases.
* With the above, it's safe to in `doDeferredStreamDestroy` access the response encoder (codec stream) and remove the envoy level stream from callbacks, and notify the codec level stream it's done.
* Removed changes to buffer account interface